fix(tianm): fix problems related to master slave synchronization (#819)

konghaiya opened the pull request:
>fix 1:#818 Master slave synchronization - There will be too many tuples problem

 Cause:
Tianmu::dbhandler::TianmuHandler::current_position This variable is not initialized. In some cases,
 a large value may cause too many tuples problem.
Solution: Initialize the variable
>fix 2:#819 Master slave synchronization - Primary key conflict problem 

Solution: Modify the modification logic of master slave synchronization, so that the delete and update operations do not follow the primary key logic

>Supplement
1. sql/sql_insert.cc 
       Fix the problem that the insert statement will not generate binlog in the delayed insert mode
2. storage/tianmu/handler/tianmu_handler.cpp 
       Fix the binlog error of the line format generated by the tianmu engine


<!--

Thank you for contributing to StoneDB!

PR Title Format: <type>(<scope>): description to this pr (#issue_id)
e.g.
fix(util): fix sth..... (#1)

-->

## Summary about this PR
<!--

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".
e.g.:
Issue: close #1

-->

Issue Number: close #818
Issue Number: close #819


## Tests Check List
<!-- At least one of next options must be included. -->

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog
<!-- At least one of next options must be included. -->

- [ ] New Feature
- [x] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation
<!-- At least one of next options must be included. -->

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

konghaiya committed:
fix(tianm) Fix problems related to master slave synchronization (#819)
fix 1:#818 Master slave synchronization - There will be too many tuples problem
Cause:
Tianmu::dbhandler::TianmuHandler::current_position This variable is not initialized. In some cases,
 a large value may cause too many tuples problem.
Solution: Initialize the variable
fix 2:#819 Master slave synchronization - Primary key conflict problem
Solution: Modify the modification logic of master slave synchronization,
so that the delete and update operations do not follow the primary key logic

RingsC commented on sql/log_event.cc:
```diff
@@ -9682,6 +9682,20 @@ search_key_in_table(TABLE *table, MY_BITMAP *bi_cols, uint key_type)
   uint res= MAX_KEY;
   uint key;
 
+  /*
```
Tests for primary-secondary scenario should be added. But, we dont see any cases.

konghaiya replied:
To add test cases

RingsC commented on sql/log_event.cc:
```diff
@@ -9682,6 +9682,20 @@ search_key_in_table(TABLE *table, MY_BITMAP *bi_cols, uint key_type)
+                      (table->s->db_type() ? table->s->db_type()->db_type == DB_TYPE_TIANMU: false);
+  enum_sql_command sqlCommand = SQLCOM_END;
+  if(table->in_use && table->in_use->lex) sqlCommand = table->in_use->lex->sql_command;
+  bool tianmuDeleteOrUpdate = (tianmu_engine && (sqlCommand == SQLCOM_DELETE ||
```
tianmuDeleteOrUpdate ？ if we dont use in the following code, it's not necssary to introduce a new variable.

RingsC replied:
The test cases should include innodb table and tianmu table both.

RingsC commented on storage/tianmu/core/pack_int.cpp:
```diff
@@ -583,19 +597,22 @@ void PackInt::LoadDataFromFile(system::Stream *fcurfile) {
       uint *cur_buf = (uint *)uptr.get();
-      if (data.ptr == nullptr && data.vt > 0) data.ptr = alloc(data.vt * dpn->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
+      if (data.ptr == nullptr && data.vt > 0)
+        data.ptr = alloc(data.vt * dpn->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
```
do memory clear after a new memory allocated.

konghaiya committed:
Merge branch 'stonedb-5.7-dev' into dev_lhj_1026

konghaiya committed:
fix(tianm): fix problems related to master slave synchronization (#819)
1. sql/sql_insert.cc Fix the problem that the insert statement will not generate binlog in the delayed insert mode
2. storage/tianmu/handler/tianmu_handler.cpp Fix the binlog error of the line format generated by the tianmu engine
